File: algorithms/279. Perfect Squares/solution.py
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        # refe to https://leetcode.com/discuss/56982/o-sqrt-n-in-ruby-and-c
        while n %  4 == 0:
            n /= 4
        if n % 8 == 7:
            return 4
        i = 0
        while i ** 2 < n:
            j = int(math.sqrt(n - i ** 2))
            if i ** 2 + j ** 2 == n:
                return bool(i) + bool(j)
            i += 1
        return 3

File: algorithms/279. Perfect Squares/solution.cc
class Solution {
public:
    int numSquares(int n) {
        if ( isSquare(n) ) {
            return 1;
        }
        while ( (n & 3) == 0 ) {
            n >>= 2;
        }
        if ( n % 8 == 7 ) {
            return 4;
        }
        for ( int i = 1; i * i < n; ++i ) {
            if ( isSquare(n - i * i) ) {
                return 2;
            }
        }
        return 3;
    }

    bool isSquare(int n) {
        int sqrtn = sqrt(n);
        return sqrtn * sqrtn == n;
    }
};
